File: M/2070-most-beautiful-item-for-each-query/main.go
package main

import (
	"fmt"
	"sort"
)

type IntArrSlice [][]int

func (p IntArrSlice) Len() int {
	return len(p)
}

func (p IntArrSlice) Swap(i, j int) {
	p[i], p[j] = p[j], p[i]
}

func (p IntArrSlice) Less(i, j int) bool {
	if p[i][0] != p[j][0] {
		return p[i][0] < p[j][0]
	}

	return p[i][1] > p[j][1]
}

func maximumBeauty(items [][]int, queries []int) []int {
	sort.Sort(IntArrSlice(items))

	base := 0
	max := -1
	buf := [][]int{}
	for i := 0; i < len(items); i++ {
		if items[i][0] != base {
			base = items[i][0]
			if max < items[i][1] {
				max = items[i][1]
			}
			buf = append(buf, []int{items[i][0], max})
		}
	}

	ret := []int{}
	for i := 0; i < len(queries); i++ {
		s, e := 0, len(buf)-1
		for s <= e {
			m := s + (e-s)/2
			if buf[m][0] > queries[i] {
				e = m - 1
			} else {
				s = m + 1
			}
		}
		if s == 0 {
			ret = append(ret, -1)
		} else {
			ret = append(ret, buf[s-1][1])
		}
	}
	return ret
}

func main() {
	fmt.Println()
}

File: M/2070-most-beautiful-item-for-each-query/main.rs
struct Solution {}

impl Solution {
  pub fn maximum_beauty(items: Vec<Vec<i32>>, queries: Vec<i32>) -> Vec<i32> {
    let mut items = items;
    items.sort_by(|x, y| x[0].cmp(&y[0]).then_with(|| x[1].cmp(&y[1])));

    let mut buf: Vec<(i32, i32)> = vec![(0, 0)];
    let mut max: i32 = 0;
    items.iter().for_each(|v| {
      let s = buf.len();
      max = max.max(v[1]);
      if v[0] > buf[s - 1].0 {
        buf.push((v[0], max));
      } else {
        buf[s - 1].1 = max;
      }
    });

    let mut ans: Vec<i32> = vec![];
    queries.iter().for_each(|&q| {
      let offset = match buf.binary_search_by(|p| p.0.cmp(&q)) {
        Ok(ov) => ov,
        Err(ev) => ev,
      };
      if offset < buf.len() && buf[offset].0 == q {
        ans.push(buf[offset].1);
      } else {
        ans.push(buf[offset - 1].1);
      }
    });
    ans
  }
}
